### DISCUSSION TITLE
Meshes from Pointwise?

### DISCUSSION THREAD

**Discussion by j-signorelli** [mfem/mfem#3585] [Q&A] [answered]
Hi,

I use Pointwise for meshing, which has a native MFEM exporter, but the actual meshes from it have always been quite incorrect. For example, here is a very simple mesh I made in Pointwise:

![image](https://user-images.githubusercontent.com/116454159/230212161-48cac16a-de12-4830-b0d8-2894d999564a.png)

After exporting it, the mesh file (with extension .mesh changed to .txt so I may paste it here) is attached: [rectangle.txt](https://github.com/mfem/mfem/files/11163155/rectangle.txt)

After running with it and outputting to Paraview (on a purely just thermal conduction simulation run), the output looks like this:

![image](https://user-images.githubusercontent.com/116454159/230213170-1055582d-f40b-48ba-9259-125ae01f0e8c.png)

Is this an issue others have dealt with before? Is there any way around this?

Thanks.

**Comment by vladotomov** [comment]
Hi @j-signorelli, I opened your file in _`GLVis`_ and right away I see the 2nd (bad) mesh. The issue seems to be at import to MFEM. Can you try even simpler, like 1 quad?

**Reply by j-signorelli** [reply]
Hi @vladotomov , thank you for the reply - I just tried a single quad, and here is the result: [single_quad.txt](https://github.com/mfem/mfem/files/11172073/single_quad.txt)

When I try to import it, I get the following error:

```
Verification failed: (faces_info[gf].Elem2No < 0) is false:
 --> Invalid mesh topology.  Interior edge found between 2D elements 0, 1 and 1.
 ... in function: void mfem::Mesh::AddSegmentFaceElement(int, int, int, int, int)
 ... in file: /home1/08335/jms26/MFEM/v4.5/mfem-4.5/mesh/mesh.cpp:6493

````

**Comment by jandrej** [comment]
Ping @dbgarlisch

**Comment by v-dobrev** [comment]
In a debug MFEM build, this mesh triggers an error:
```
Assertion failed: (i>=0 && i<size) is false:
 --> Access element 2500 of array, size = 2500
 ... in function: T &mfem::Array<int>::operator[](int) [T = int]
 ... in file: general/array.hpp:743

Backtrace:
0) [0x7c2b00010419281b]: mfem::mfem_error(char const*)
1) [0xcf5c0001041797d3]: mfem::Array<int>::operator[](int)
2) [0x701780010426e2ab]: mfem::Mesh::RemoveUnusedVertices()
3) [0x2f7d8001042a2eaf]: mfem::Mesh::ReadMFEMMesh(std::__1::basic_istream<char, std::__1::char_traits<char>>&, int, int&)
4) [0x2e37800104268167]: mfem::Mesh::Loader(std::__1::basic_istream<char, std::__1::char_traits<char>>&, int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>)
5) [0x330580010426327f]: mfem::Mesh::Load(std::__1::basic_istream<char, std::__1::char_traits<char>>&, int, int, bool)
6) [0x3237800104265cbf]: mfem::Mesh::Mesh(char const*, int, int, bool)
7) [0xad46000104265f7f]: mfem::Mesh::Mesh(char const*, int, int, bool)
8) [0xd66700010415d973]: main

Lookup backtrace source lines:
atos -o /<truncated>/mfem/miniapps/meshing/mesh-explorer -l 0x104158000 0x7c2b00010419281b 0xcf5c0001041797d3 0x701780010426e2ab 0x2f7d8001042a2eaf 0x2e37800104268167 0x330580010426327f 0x3237800104265cbf 0xad46000104265f7f 0xd66700010415d973

Abort
```
The source code locations from the backtrace are:
```
mfem::mfem_error(char const*) (in mesh-explorer) (error.cpp:167)
mfem::Array<int>::operator[](int) (in mesh-explorer) (array.hpp:742)
mfem::Mesh::RemoveUnusedVertices() (in mesh-explorer) (mesh.cpp:11790)
mfem::Mesh::ReadMFEMMesh(std::__1::basic_istream<char, std::__1::char_traits<char>>&, int, int&) (in mesh-explorer) (mesh_readers.cpp:105)
mfem::Mesh::Loader(std::__1::basic_istream<char, std::__1::char_traits<char>>&, int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>) (in mesh-explorer) (mesh.cpp:4053)
mfem::Mesh::Load(std::__1::basic_istream<char, std::__1::char_traits<char>>&, int, int, bool) (in mesh-explorer) (mesh.hpp:917)
mfem::Mesh::Mesh(char const*, int, int, bool) (in mesh-explorer) (mesh.cpp:3812)
mfem::Mesh::Mesh(char const*, int, int, bool) (in mesh-explorer) (mesh.cpp:3800)
main (in mesh-explorer) (mesh-explorer.cpp:306)
```
This means that there are some elements in the mesh that use vertex indices larger than the declared number of vertices -- specifically some elements uses index 2500 and there are 2500 vertices.

**Comment by v-dobrev** [comment, marked as answer]
Looking at the mesh file, I suspect that the vertex indices describing the elements and the boundary need to be shifted down by 1 to be in the range 0-2499 instead of 1-2500.

**Reply by dbgarlisch** [reply]
I will look into this in the pointwise code.
You can also call pw support and log a bug.

**Reply by j-signorelli** [reply]
Thanks so much for looking into this, I will log this with pw support. In the meantime I'll write a script that edits pw exported files as stated above and see if that works.

**Reply by j-signorelli** [reply]
This is correct, updating the vertex indices using a Python script seems to have resolved this issue

**Reply by dbgarlisch** [reply]
I could not find anywhere in the MFEM format doc explicitly stating that it uses 0-based vertex indexing.

However, the [one doc example][1] given for a linear (straight) 2-D mesh ([beam-quad.mesh][2]) shows elements and boundaries including vertex indices with value 0.

I have confirmed that the Pointwise MFEM plugin code is incorrectly exporting the element connectivity using 1-based indices. It will be fixed.

[1]: https://mfem.org/mesh-format-v1.0/#mfem-mesh-v10
[2]: https://github.com/mfem/mfem/blob/master/data/beam-quad.mesh

**Reply by dbgarlisch** [reply]
Logged [MFEM Issue 3595](https://github.com/mfem/mfem/issues/3595) to clarify documentation
